File: py/behavioral/classic_iterator.py
# Iterator lets clients traverse a collection without exposing how it stores items.
# Modern Python note: iteration is language-native. Most code should use iterables,
# generators, or __iter__ rather than custom iterator class hierarchies.
class Playlist:
    def __init__(self, songs):
        self.songs = songs

    def __iter__(self):
        return iter(self.songs)


if __name__ == "__main__":
    playlist = Playlist(["Intro", "Main Theme", "Finale"])
    for song in playlist:
        print(song)

File: csharp/behavioral/classic_iterator.cs
namespace Gof.ModernCSharp.Behavioral;

// Iterator lets clients traverse a collection without exposing how it stores items.
// Modern C# note: iteration is language-native through IEnumerable<T>, yield
// return, LINQ, and foreach. Custom iterator classes are usually unnecessary.
public static class ClassicIterator
{
    sealed class Playlist(IEnumerable<string> songs) : IEnumerable<string>
    {
        public IEnumerator<string> GetEnumerator() => songs.GetEnumerator();
        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
    }

    public static string Run() => string.Join(" > ", new Playlist(["Intro", "Main Theme", "Finale"]));
}
